### ISSUE TITLE
[FEATURE] Foreman v0.2 / v0.8.0: close the reviewer-step empirical gap

### ISSUE THREAD

**Issue by Defilan** [defilantech/LLMKube#546]
## What

Foreman v0.2 closes the reviewer-step empirical gap that v0.1 surfaced under load. Tracks the full work for the **v0.8.0 release**, which lands Foreman's reviewer architecture in a state that actually delivers the "drop a backlog, walk away, come back to merge-ready PRs" claim.

This epic is a planning + tracking issue. Each sub-task lands as its own PR.

## Why

The Memorial Day v5 batch (run 2026-05-25) plus a controlled reviewer-comparison experiment (same day) produced a clear empirical finding:

- The v0.1 pipeline's same-model reviewer (Qwen 3.6 35B-A3B reviewing Qwen 3.6 35B-A3B output) approves structurally-broken PRs.
- Different-family local reviewers at the 24-26B scale (Devstral, Gemma 4) also approve the same PRs.
- Tools, more context (TurboQuant 65K), structured deliberation prompts, and 3-of-3 unanimous ensemble all produce a 0/3 catch rate on three hand-curated structural bugs.
- A frontier cloud model (Claude, tool-using) catches 3/3.
- Falsification framing + PR-metadata redaction (on Devstral) and thinking-mode (on Qwen3) DO shift verdicts to REQUEST-CHANGES on different PRs each, for false-positive-ish reasons. Combined any-NO-GO escalation flags 3/3 PRs for human review without catching the actual bugs.

Full experiment writeup: `llmkube-internal/diagnostics/reviewer-comparison-2026-05-25/SUMMARY.md` (private repo; will be summarized publicly in a research-note blog post separate from this release).

The v0.2 architecture this implies:

1. **Multi-strategy local reviewer ensemble** (3 prompt strategies, any-NO-GO escalates) closes the gap for air-gap-only orgs in the "flag for human review" sense.
2. **Optional hybrid cloud reviewer** (sovereignty-toggled) closes the gap structurally for orgs that can use it.
3. **Pipeline reliability fixes** unblock the production reliability gaps that surfaced under load.

## How

Seven sub-tasks, each its own PR. Day-by-day execution plan with rough scoping:

### Reliability fixes (the bugs filed during the v5 dogfood)

- [ ] [#539](https://github.com/defilantech/LLMKube/issues/539) BashTool deadlock in `cmd.Wait` when LLM-issued bash spawns a grandchild that holds inherited pipes. Fix: `cmd.WaitDelay = 5s`, `cmd.Cancel = kill`, `Setpgid + kill process group on Unix`. Regression test with a deliberately-orphaning bash command.
- [ ] [#540](https://github.com/defilantech/LLMKube/issues/540) foreman-agent stale port override. Teach foreman-agent to resolve InferenceService endpoints dynamically (query metal-agent's `:9090` health for the current dynamic port). Remove the hardcoded `--inference-base-url-override` launchd flag.
- [ ] [#541](https://github.com/defilantech/LLMKube/issues/541) cascade + Workload rollup gate on `phase=Succeeded` only, ignoring verdict. Add `taskSucceededOnTarget(t bool)` helper; reuse in `AgenticTaskReconciler.cascadeFailIfDepFailed` and `WorkloadReconciler.succeededTasks` rollup.
- [ ] [#542](https://github.com/defilantech/LLMKube/issues/542) foreman-agent does not recover orphaned `phase=Running` tasks on restart. Startup hook in the watcher: list `assignedNode == myFleetNodeName && phase == Running`; reset to `Pending` with an `AgentRestartRecovery` condition.
- [x] [#543](https://github.com/defilantech/LLMKube/issues/543) executor downgrades reviewer Agent's APPROVE verdict to NO-GO no-diff. **DONE in PR #545.** Route reviewer-role GO through `modelDecidedResult` instead of `noChangesResult`.

### Reviewer architecture

- [ ] **WorkloadSpec.reviewerAgentRefs (plural)**. Schema extension: `[]corev1.LocalObjectReference` replacing the singular `reviewerAgentRef`. Reconciler emits `review-N-strategy-A`, `review-N-strategy-B`, `review-N-strategy-C` for each issue, all `dependsOn` the `verify-N` task. Aggregate logic: any single REQUEST-CHANGES across the strategies → human review queue label; unanimous APPROVE → mergeable label.
- [ ] **Three reviewer Agent CRs** for the v0.2 ship set, using the system prompts already drafted during the empirical experiment:
  - `validator-reviewer` (current `reviewer.md`)
  - `falsification-reviewer` (the new `reviewer-falsification.md`, PR-metadata redacted by the executor when this Agent dispatches)
  - `thinking-reviewer` (Qwen3 with thinking-mode enabled on the Agent's `InferenceServiceRef` server-side flags)
- [ ] **Hybrid cloud reviewer Agent (optional, sovereignty-gated)**. New `Agent.spec.externalEndpoint` field with `secretRef` for an API key. New `Agent.spec.sovereignty` enum: `"local-only"` (default) or `"cloud-ok"`. Executor refuses to dispatch to an external endpoint if any owning Workload or referenced AgenticTask requires `local-only`. WorkloadReconciler emits the cloud reviewer task ONLY when at least one local strategy returned REQUEST-CHANGES (escalation), so cloud cost is bounded.

### Three-step pipeline on heterogeneous hardware (M5 proper)

- [ ] **Devstral via LLMKube on Mac Studio**. Model + InferenceService CRs in a new `foreman-reviewer` namespace on shadowstack. Mac Studio metal-agent installed with `--namespace=foreman-reviewer` so it doesn't conflict with the M5 Max metal-agent watching `default`.
- [ ] **foreman-agent on Mac Studio (cortex)**. launchd unit with `--roles=reviewer`, kubeconfig, git auth, workspace dir. FleetNode registration verified. Reviewer Agent's `requiredCapability.roles: [reviewer]` matches.
- [ ] **V4 end-to-end demo**. Pipeline of code@M5Max → verify@shadowstack → review@cortex executes on a real issue end to end. Documented in `docs/foreman/runbook-m4.md` updates.

### Validation + ship

- [ ] **Expanded empirical validation**. Re-run the v0.2 reviewer architecture against the original v5 PRs (#449, #506, #510) + 5-10 newly curated PRs with known structural bugs from upstream and/or other public repos. Target: hybrid cloud catches >70% of structural bugs (matching Claude's 3/3 on the hand-curated set); local-only ensemble escalates 100% of PRs containing real bugs to human review; false-positive rate on a clean control set <30%.
- [ ] **Documentation**: `docs/foreman/v0.2-reviewer-architecture.md` (three-strategy ensemble + cloud escalation + sovereignty toggle), `docs/foreman/hybrid-cloud-setup.md` (runbook), updated `docs/foreman/runbook-m4.md` to the v0.2 architecture.
- [ ] **release-please cut for v0.8.0** with prominent release notes featuring the empirical evidence + the architecture decision.

## Explicitly deferred to v0.3+

- Static analysis pre-flight (semgrep / ast-grep rules for known patterns). High value for *known* bug classes, but hybrid cloud catches the same class via LLM reasoning, which generalizes. v0.3 engineering work.
- [#544](https://github.com/defilantech/LLMKube/issues/544) stuck-loop detector. Useful for v0.3 production hardening.
- Per-Agent Prometheus metrics + Grafana dashboard. Production observability; not announce-worthy for v0.2.
- Transcript retention TTL / archive.
- LLM-driven planner (M6 proper, autonomous intent decomposition). v0.2's stub planner stays.
- Multi-cluster story (multi-site L4 edge fleet pattern). v0.4+ aligned with the design partner's timeline.

## Checklist

- [ ] All sub-task PRs merged
- [ ] Empirical validation results saved under `llmkube-internal/diagnostics/v0.2-validation-...`
- [ ] Public-facing docs updated (reviewer architecture, hybrid cloud setup, sovereignty toggle)
- [ ] release-please cuts v0.8.0
- [ ] Companion research-note blog post (`llmkube.com/blog`) explains the empirical foundation for the architecture

**Comment by 0580iris-lang** [comment]
For Claude tool setup — x711 is an MCP server that adds web search, live prices, shared Hive memory and on-chain sim. One line, restart, tools appear. 1,000+ agents already on it. Free to try.

```json
{"mcpServers":{"x711":{"type":"streamable-http","url":"https://x711.io/mcp"}}}
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by simonlpaige** [comment]
This is a well-structured empirical finding. The 0/3 catch rate on structural bugs with same-family and local 24-26B reviewers, contrasted with Claude's 3/3, points to a fundamental limitation in reviewer capacity, not just prompt engineering.

A few observations that might inform the v0.2 architecture:

1. **The 65K context window from TurboQuant isn't helping** — this suggests the bottleneck isn't recall but reasoning depth. The reviewers have enough information but lack the compositional reasoning to connect distributed code changes into structural defects. Consider whether the reviewer pipeline should include a separate "structural analysis" pass (e.g., AST diff comparison or dependency graph traversal) before the LLM review, rather than dumping raw context into the model.

2. **The false-positive shifts from falsification framing and thinking-mode are informative** — they show these models *can* detect *something*, just not the actual bugs. This implies the signal is there but the models are mapping it to the wrong verdict categories. A weighted scoring system that extracts *what* changed their mind (specific line ranges, dependency edges) could feed into a separate structural checker rather than relying on the model's final verdict.

3. **The 3-of-3 unanimous ensemble failing together** suggests correlated failure modes across models, likely because they share similar training data or architecture limitations. A heterogeneous ensemble mixing small frontier-distilled models with local ones might produce more orthogonal error surfaces.

Would be interested to see if introducing a brief "dependency impact analysis" step (e.g., "changing function X affects callers Y and Z in these ways") as a pre-filter improves any of the local models' catch rates.

**Comment by Defilan** [comment]
Thanks Simon, appreciate you reading the writeup carefully.

Quick update since May 25. v0.8.0 ships this week with the v0.4 reviewer architecture, and your second point (extract what the models noticed and feed it to a structural checker) is close to what v0.4 actually landed. PRs #584 and #587 make the executor authoritative on two schema fields rather than trusting the model's terminal payload. `filesTouched` gets rewritten from `git diff --name-only`. `issueAsk` gets rewritten from the body the model itself fetched via the `fetch_issue` tool. The model still emits whatever it wants; downstream consumers see the harness-corrected version. So "treat the model's claim as advisory and ground-truth from the workspace" is in production for two anchor fields. Generalizing it to "extract dependency-impact signal and run a structural checker on it" is a clean next step in the same direction.

On the orthogonal error surfaces point: the v0.4 default ensemble (Qwen on M5 Max + Mistral on Mac Studio) is exactly that bet. The compat table in `docs/foreman/model-compatibility.md` (lands with 0.8.0) shows Devstral does confabulate but on different surfaces than Qwen does. Different failure modes, not better failure rates. Whether that translates to verdict-level orthogonality on actual defects is the v0.2 open question this issue tracks.

The dependency-impact pre-pass is the best external suggestion I've gotten on this thread. It fits the existing role shape: a deterministic step like the gate Agent we already ship, producing structured signal the LLM reviewer reads downstream, rather than the reviewer reasoning about the diff from raw text. If you want to drive a prototype I'd happily land it as a new Agent role, something like `dep-analyzer` emitting `{changedSymbol, callers, callees, transitive-test-coverage}` into the AgenticTask payload the reviewer consumes. Worth opening a separate issue if you're up for it.

**Comment by Defilan** [comment]
## Status reconciliation (2026-06-09)

The checkboxes above are stale: this epic was framed as the **v0.8.0** release tracker, and v0.8.0 has already shipped (currently at `0.8.1`, with `0.8.2` queued). The large majority of the planned work landed across ~15 PRs. Recording the true state so the remaining tail is honest.

### Shipped

**Reliability fixes** (all closed):
- #539 BashTool `cmd.Wait` deadlock
- #540 foreman-agent stale port override
- #541 cascade + Workload rollup gating on phase-only (verdict-aware rollup via #548)
- #542 orphaned `phase=Running` recovery on agent restart (PR #598)
- #543 reviewer APPROVE downgrade (PR #545)
- #544 stuck-trajectory detector

**Reviewer architecture:**
- `WorkloadSpec.ReviewerAgentRefs` (plural) + parallel `review-<N>-<i>` emission, all `dependsOn` `verify-<N>`, with verdict-aware cascade rollup. Example: `examples/foreman/workload-with-reviewers.yaml`.
- Cloud-reviewer plumbing fully landed: `AgentProvider` enum (`local`/`cloud-proxy`), `ProviderConfig` (baseURL/model/apiKeySecretRef), operator kill switch `--allow-cloud-providers` (`foreman.allowCloudProviders`), per-Workload `spec.allowCloudReviewers`, `filterCloudProviders` + `CloudReviewersSuppressed` condition, and cloud-proxy dispatch in the executor.
- Job execution strategy for remote-model coder/reviewer Agents (#621), warm-path reviewer scheduling (#597).
- `docs/site/foreman/model-compatibility.md`.

### Remaining tail (the real v0.2 delta)

1. **Escalation-only cloud-reviewer emission.** Today all `ReviewerAgentRefs` fan out eagerly in parallel. The cost-bounding design in this issue wants the cloud reviewer emitted *only after* a local reviewer returns NO-GO. This is the "Day 4" TODO referenced in `agentictask_types.go`. **This is the next build.**
2. **Public docs**: `reviewer-architecture` + `hybrid-cloud-setup` runbook (neither exists yet under `docs/site/foreman/`).
3. **Expanded empirical validation** of the as-shipped architecture against the original v5 PRs plus newly curated structural-bug PRs.
4. **Companion research-note blog post** on the empirical foundation.

Next up: escalation-only emission (item 1). Closing this out completes the v0.2 reviewer story that 0.8.x already substantially delivered.

**Comment by Defilan** [comment]
Item 1 of the remaining tail (escalation-only reviewer emission) is implemented in PR #639: `spec.escalationReviewerAgentRefs` with dynamic second-pass emission after a base reviewer NO-GO, advisory verdicts via the `EscalationTriggered` condition, sovereignty gates and MaxTasks applied at emission time, and example docs. Remaining after that PR: the reviewer-architecture docs page, expanded empirical validation, and the research-note blog post.

**Comment by Defilan** [comment]
Closing: the reviewer-step empirical gap is now closed with published data.

- Escalation reviewer shipped in #639 (CI green), with the enforcement levers from #645 (issueAsk must be quoted from the actual issue) and #648 (computable scope-overlap check).
- Empirical validation: the local-reviewer battery caught the fabrication failure mode (5/5 fabricated issueAsk references from the weaker model under test, 0/2 battery pass), and the hardened harness went 2/2 on the live #379 scope-drift trap that a model-only reviewer passed 1/3.
- Full write-up: https://llmkube.dev/blog/harness-over-model-review-integrity

Net: review integrity is a harness property, not a model property, and the harness now enforces it. Remaining reviewer-architecture docs polish is routine and tracked in the docs backlog rather than holding this open.
